File: app/src/main/java/com/cn29/aac/binding/TextInputEditTextBindingUtil.java
package com.cn29.aac.binding;

import android.databinding.BindingAdapter;
import android.support.design.widget.TextInputLayout;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;

/**
 * Created by Charles Ng on 7/9/2017.
 */

public class TextInputEditTextBindingUtil {


  @BindingAdapter({"validation", "errorMsg"})
  public static void setErrorEnable(TextInputLayout textInputLayout, StringRule stringRule,
      final String errorMsg) {
    textInputLayout.getEditText().addTextChangedListener(new TextWatcher() {
      @Override
      public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
      }

      @Override
      public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
      }

      @Override
      public void afterTextChanged(Editable editable) {
        textInputLayout
            .setErrorEnabled(stringRule.validate(textInputLayout.getEditText().getText()));
        if (stringRule.validate(textInputLayout.getEditText().getText())) {
          textInputLayout.setError(errorMsg);
        } else {
          textInputLayout.setError(null);
        }
      }
    });
    textInputLayout
        .setErrorEnabled(stringRule.validate(textInputLayout.getEditText().getText()));
    if (stringRule.validate(textInputLayout.getEditText().getText())) {
      textInputLayout.setError(errorMsg);
    } else {
      textInputLayout.setError(null);
    }
  }

  public interface StringRule {

    boolean validate(Editable s);
  }

  public static class Rule {

    public static StringRule NOT_EMPTY_RULE = s -> TextUtils.isEmpty(s.toString());
    public static StringRule EMAIL_RULE = s -> s.toString().length() > 18;
  }

}

File: app/src/main/java/com/cn29/aac/binding/TextInputEditTextBindingUtil.kt
package com.cn29.aac.binding

import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import androidx.databinding.BindingAdapter
import com.google.android.material.textfield.TextInputLayout

/**
 * Created by Charles Ng on 7/9/2017.
 */
object TextInputEditTextBindingUtil {
    @JvmStatic
    @BindingAdapter("validation", "errorMsg")
    fun setErrorEnable(textInputLayout: TextInputLayout,
                       stringRule: StringRule,
                       errorMsg: String?) {
        textInputLayout.editText
                ?.addTextChangedListener(object : TextWatcher {
                    override fun beforeTextChanged(charSequence: CharSequence,
                                                   i: Int,
                                                   i1: Int,
                                                   i2: Int) {
                    }

                    override fun onTextChanged(charSequence: CharSequence,
                                               i: Int,
                                               i1: Int,
                                               i2: Int) {
                    }

                    override fun afterTextChanged(editable: Editable) {
                        textInputLayout.isErrorEnabled = stringRule.validate(
                                textInputLayout.editText
                                        ?.text)
                        if (stringRule.validate(textInputLayout.editText
                                                        ?.text)) {
                            textInputLayout.error = errorMsg
                        } else {
                            textInputLayout.error = null
                        }
                    }
                })
        textInputLayout.isErrorEnabled = stringRule.validate(textInputLayout.editText
                                                                     ?.text)
        if (stringRule.validate(textInputLayout.editText!!.text)) {
            textInputLayout.error = errorMsg
        } else {
            textInputLayout.error = null
        }
    }

    interface StringRule {
        fun validate(s: Editable?): Boolean
    }

    object Rule {
        @JvmField
        var NOT_EMPTY_RULE: StringRule = object : StringRule {
            override fun validate(s: Editable?): Boolean {
                return TextUtils.isEmpty(s.toString())
            }
        }
        var EMAIL_RULE: StringRule = object : StringRule {
            override fun validate(s: Editable?): Boolean {
                return s.toString().length > 18
            }
        }
    }
}